jerry-core/debugger/debugger.c

```diff
@@ -195,6 +196,333 @@ jerry_debugger_send_backtrace (const uint8_t *recv_buffer_p) /**< pointer to the
   jerry_debugger_send (sizeof (jerry_debugger_send_type_t) + message_size);
 } /* jerry_debugger_send_backtrace */
 
+
+/**
+ * Send the scope chain types.
+ */
+static void
+jerry_debugger_send_scope_chain (void)
+{
+  vm_frame_ctx_t *iter_frame_ctx_p = JERRY_CONTEXT (vm_top_context_p);
+
+  const size_t max_byte_count = JERRY_DEBUGGER_SEND_MAX (uint8_t);
+  const size_t max_message_size = JERRY_DEBUGGER_SEND_SIZE (max_byte_count, uint8_t);
+
+  JERRY_DEBUGGER_SEND_BUFFER_AS (jerry_debugger_send_string_t, message_type_p);
+  message_type_p->type = JERRY_DEBUGGER_SCOPE_CHAIN;
+
+  size_t buffer_pos = 0;
+  uint8_t level_type;
+  bool next_func_is_local = true;
+
+  for (ecma_object_t *lex_env_p = iter_frame_ctx_p->lex_env_p;
+       lex_env_p != NULL;
+       lex_env_p = ecma_get_lex_env_outer_reference (lex_env_p))
+  {
+    JERRY_ASSERT (ecma_is_lexical_environment (lex_env_p));
+
+    if (ecma_get_lex_env_type (lex_env_p) == ECMA_LEXICAL_ENVIRONMENT_DECLARATIVE)
+    {
+      if ((lex_env_p->type_flags_refs & ECMA_OBJECT_FLAG_EXTENSIBLE) != 0)
+      {
+        level_type = JERRY_DEBUGGER_SCOPE_CATCH;
+      }
+      else if (next_func_is_local)
+      {
+        level_type = JERRY_DEBUGGER_SCOPE_LOCAL;
+        next_func_is_local = false;
+      }
+      else
+      {
+        level_type = JERRY_DEBUGGER_SCOPE_CLOSURE;
+        next_func_is_local = false;
+      }
+    }
+    else
+    {
+      JERRY_ASSERT (ecma_get_lex_env_type (lex_env_p) == ECMA_LEXICAL_ENVIRONMENT_THIS_OBJECT_BOUND);
+
+      if (ecma_get_lex_env_outer_reference (lex_env_p) == NULL)
+      {
+        level_type = JERRY_DEBUGGER_SCOPE_GLOBAL;
+      }
+      else
+      {
+        level_type = JERRY_DEBUGGER_SCOPE_WITH;
+        next_func_is_local = true;
+      }
+    }
+
+    message_type_p->string[buffer_pos++] = level_type;
+
+    if (buffer_pos == max_byte_count)
+    {
+      if (!jerry_debugger_send (max_message_size))
+      {
+        return;
+      }
+
+      buffer_pos = 0;
+    }
+  }
+
+  message_type_p->type = JERRY_DEBUGGER_SCOPE_CHAIN_END;
+
+  jerry_debugger_send (sizeof (jerry_debugger_send_type_t) + buffer_pos);
+} /* jerry_debugger_send_scope_chain */
+
+/**
+ * Get type of the scope variable property.
+ */
+static jerry_debugger_scope_variable_type_t
+ecma_get_typeof_scope_variable (ecma_value_t value) /**< input ecma value */
+{
+  jerry_debugger_scope_variable_type_t ret_value = JERRY_DEBUGGER_VARIABLE_NONE;
+
+  if (ecma_is_value_undefined (value))
+  {
+    ret_value = JERRY_DEBUGGER_VARIABLE_UNDEFINED;
```

Comment:
It might be nitpicking, but I think VALUE is better than VARIABLE, since the value of a variable is undefined, rather than the variable, which would mean it does not exist.